### PR TITLE
Text replacement animations do not run in the correct order and fail to animate over the correct ranges when text comes in over time.

### PR DESCRIPTION
**Pull request by megangardner**
#### 5c25cad683ccc1832ce4f12f283a032a2be9f4ac
<pre>
Text replacement animations do not run in the correct order and fail to animate over the correct ranges when text comes in over time.
<a href="https://bugs.webkit.org/show_bug.cgi?id=276113">https://bugs.webkit.org/show_bug.cgi?id=276113</a>
<a href="https://rdar.apple.com/128292335">rdar://128292335</a>

Reviewed by Aditya Keerthi and Wenson Hsieh.

I was not using the full async animation infrastructure
and the tracking of the different ranges as the different
chunks came in was not correct. These changes should
streamline the animations and correct some not-as-good names
for some of the variables. There are still some frames that
need a bit of polish, but this is a much better working version
than we have had up to this point.

* Source/WebCore/editing/CharacterRange.h:
(WebCore::CharacterRange::operator== const):
* Source/WebCore/page/ChromeClient.h:
(WebCore::ChromeClient::addSourceTextAnimation):
* Source/WebCore/page/writing-tools/WritingToolsController.h:
* Source/WebCore/page/writing-tools/WritingToolsController.mm:
(WebCore::WritingToolsController::compositionSessionDidReceiveTextWithReplacementRange):
* Source/WebKit/Shared/TextAnimationType.serialization.in:
* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
(WebKit::WebPageProxy::addTextAnimationForAnimationID):
(WebKit::WebPageProxy::callCompletionHandlerForAnimationID):
(WebKit::WebPageProxy::getTextIndicatorForID):
* Source/WebKit/UIProcess/TextAnimationType.h:
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/UIProcess/WebPageProxyInternals.h:
* Source/WebKit/UIProcess/mac/WKTextAnimationManager.mm:
(-[WKTextAnimationManager addTextAnimationForAnimationID:withData:]):
(-[WKTextAnimationManager textPreviewsForChunk:completion:]):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::addSourceTextAnimation):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
* Source/WebKit/WebProcess/WebPage/Cocoa/TextAnimationController.h:
* Source/WebKit/WebProcess/WebPage/Cocoa/TextAnimationController.mm:
(WebKit::TextAnimationController::unreplacedRangeForSessionWithID const):
(WebKit::TextAnimationController::contextRangeForTextAnimationID const):
(WebKit::TextAnimationController::removeTransparentMarkersForSessionID):
(WebKit::remainingCharacterRange):
(WebKit::TextAnimationController::addInitialTextAnimation):
(WebKit::TextAnimationController::addSourceTextAnimation):
(WebKit::TextAnimationController::addDestinationTextAnimation):
(WebKit::TextAnimationController::updateUnderlyingTextVisibilityForTextAnimationID):
(WebKit::newlyReplacedCharacterRange): Deleted.
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::addTextAnimationForAnimationID):
(WebKit::WebPage::addSourceTextAnimation):
* Source/WebKit/WebProcess/WebPage/WebPage.h:

Canonical link: <a href="https://commits.webkit.org/280763@main">https://commits.webkit.org/280763@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/bf31a3bfdb680bdcd10668242db17db0d6465895

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/57514 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/36842 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/9989 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/61136 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/7959 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/59642 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/44475 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/8147 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/46574 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/5640 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/59544 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/34574 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/49682 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/27438 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/31357 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/6985 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/6962 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/53320 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/7257 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/62815 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/1427 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/7350 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/53832 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/1434 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/49707 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/53951 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12725 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/1218 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/32671 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/33756 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/34841 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/33502 "Built successfully") | | |

### CODE REVIEW

File: Source/WebCore/editing/CharacterRange.h
Comment (pxlcoder) [inline Source/WebCore/editing/CharacterRange.h]: I think you can use `= default` here.

File: Source/WebCore/page/writing-tools/WritingToolsController.mm
Comment (pxlcoder) [inline Source/WebCore/page/writing-tools/WritingToolsController.mm]: `WritingToolsController` already has a weak reference to page itself, so you don't need `m_page` in the capture list.

Comment (pxlcoder) [inline Source/WebCore/page/writing-tools/WritingToolsController.mm]: `movedFragement` (typo) can just be called `fragment`.

File: Source/WebKit/UIProcess/mac/WKTextAnimationManager.mm
Comment (pxlcoder) [inline Source/WebKit/UIProcess/mac/WKTextAnimationManager.mm]: Unnecessary whitespace.

File: Source/WebKit/WebProcess/WebPage/Cocoa/TextAnimationController.mm
Comment (pxlcoder) [inline Source/WebKit/WebProcess/WebPage/Cocoa/TextAnimationController.mm]:
```suggestion
        auto animationRange = contextRangeForTextAnimationID(uuid);
```

File: Source/WebKit/UIProcess/mac/WKTextAnimationManager.mm
Comment (pxlcoder) [inline Source/WebKit/UIProcess/mac/WKTextAnimationManager.mm]:
```suggestion
            effect.get().preCompletion = makeBlockPtr([weakWebView = WeakPtr<WebKit::WebViewImpl>(_webView), uuid = retainPtr(uuid)] {
```

I think you need to retain, since nothing else guarantees the lifetime of `NSUUID *`?

File: Source/WebKit/UIProcess/mac/WKTextAnimationManager.mm
Comment (pxlcoder) [inline Source/WebKit/UIProcess/mac/WKTextAnimationManager.mm]:
```suggestion
                auto animationID = WTF::UUID::fromNSUUID(uuid.get());
```

File: Source/WebCore/editing/CharacterRange.h
Comment (megangardner) [reply Source/WebCore/editing/CharacterRange.h]: I have no idea how to format that, and searching our codebase isn't helping...

Comment (pxlcoder) [reply Source/WebCore/editing/CharacterRange.h]: `bool operator==(const CharacterRange& range) const = default;`

File: Source/WebCore/page/writing-tools/WritingToolsController.mm
Comment (rr-codes) [inline Source/WebCore/page/writing-tools/WritingToolsController.mm]: where is `fragment` or `movedFragment` used?

File: Source/WebCore/page/writing-tools/WritingToolsController.mm
Comment (rr-codes) [inline Source/WebCore/page/writing-tools/WritingToolsController.mm:432]: nit: a newline before the `replaceContentsOfRangeInSession` call would be nice imo

File: Source/WebCore/page/writing-tools/WritingToolsController.mm
Comment (rr-codes) [inline Source/WebCore/page/writing-tools/WritingToolsController.mm:439]: this should also have an `ASSERT_NOT_REACHED` before the return

File: Source/WebKit/UIProcess/mac/WKTextAnimationManager.mm
Comment (rr-codes) [inline Source/WebKit/UIProcess/mac/WKTextAnimationManager.mm]: Assuming this respondsToSelector was originally adding for staging purposes, it can probably be removed now, right?

File: Source/WebCore/page/writing-tools/WritingToolsController.mm
Comment (megangardner) [reply Source/WebCore/page/writing-tools/WritingToolsController.mm:439]: We actually run into this more than I would think, and we need to not be calculating this in this manner anyways, so I'm going to leave it for now.

File: Source/WebCore/page/writing-tools/WritingToolsController.mm
Comment (pxlcoder) [inline Source/WebCore/page/writing-tools/WritingToolsController.mm]: Nit: Would remove this newline.

File: Source/WebKit/UIProcess/mac/WKTextAnimationManager.mm
Comment (pxlcoder) [inline Source/WebKit/UIProcess/mac/WKTextAnimationManager.mm]: Maybe just `animationID`? `wtf` in the variable is name is weird 😅

File: Source/WebKit/UIProcess/mac/WKTextAnimationManager.mm
Comment (pxlcoder) [inline Source/WebKit/UIProcess/mac/WKTextAnimationManager.mm]: Same comment as above.

File: Source/WebKit/WebProcess/WebPage/Cocoa/TextAnimationController.mm
Comment (pxlcoder) [inline Source/WebKit/WebProcess/WebPage/Cocoa/TextAnimationController.mm]:
```suggestion
    // Take out the part that has already been replaced.
```

If we're keeping this comment...

File: Source/WebKit/WebProcess/WebPage/Cocoa/TextAnimationController.mm
Comment (pxlcoder) [inline Source/WebKit/WebProcess/WebPage/Cocoa/TextAnimationController.mm]: Nit: Remove newline.

File: Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm
Comment (pxlcoder) [inline Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:1252]: If the web process dies, I think `completionHandlerForAnimationID` is going to be left with stale completion handlers.

`WebPageProxy::resetState` should probably clear the map.

File: Source/WebCore/page/writing-tools/WritingToolsController.mm
Comment (pxlcoder) [inline Source/WebCore/page/writing-tools/WritingToolsController.mm]: `m_page` should probably be null checked.

File: Source/WebKit/UIProcess/WebPageProxyInternals.h
Comment (pxlcoder) [inline Source/WebKit/UIProcess/WebPageProxyInternals.h:298]: Asked a question to the team about whether this should be `Function` rather than `CompletionHandler`.

**Review by pxlcoder — APPROVED** [review]
Please amend the commit message to remove the title next to the radar.

File: Source/WebKit/WebProcess/WebPage/Cocoa/TextAnimationController.mm
Comment (whsieh) [inline Source/WebKit/WebProcess/WebPage/Cocoa/TextAnimationController.mm]: (Please remove this debug logging before landing!)

Comment (pxlcoder) [reply Source/WebKit/WebProcess/WebPage/Cocoa/TextAnimationController.mm]: That was not there in the version of the patch I reviewed!

File: Source/WebKit/UIProcess/WebPageProxyInternals.h
Comment (pxlcoder) [reply Source/WebKit/UIProcess/WebPageProxyInternals.h:298]: We are leaving as `CompletionHandler` since it's expected to be called in all cases except where the web process is killed.

File: Source/WebKit/WebProcess/WebPage/Cocoa/TextAnimationController.mm
Comment (whsieh) [inline Source/WebKit/WebProcess/WebPage/Cocoa/TextAnimationController.mm]: It's a bit weird for WebKit (client) code to need to call into the `Chrome`/`ChromeClient` just to propagate back up to the client layer.

Maybe simpler if you just call into `WebPage` directly?

```suggestion
        weakWebPage->addInitialTextAnimation(sessionUUID);
```

File: Source/WebKit/WebKitSwift/TextAnimation/TextAnimationManager.swift
Comment (whsieh) [inline Source/WebKit/WebKitSwift/TextAnimation/TextAnimationManager.swift]: Nit - extra space before the `)`

```suggestion
                    Self.logger.debug("Can't obtain Targeted Preview. Missing delegate.")
```

**Review by whsieh — APPROVED** [review]

File: Source/WebKit/WebProcess/WebPage/Cocoa/TextAnimationController.mm
Comment (megangardner) [reply Source/WebKit/WebProcess/WebPage/Cocoa/TextAnimationController.mm]: This sneaked in with the stuff I did to make sure iOS was working.

File: Source/WebKit/UIProcess/mac/WKTextAnimationManager.mm
Comment (pxlcoder) [inline Source/WebKit/UIProcess/mac/WKTextAnimationManager.mm]:
```suggestion
        effect.get().completion = makeBlockPtr([weakWebView = WeakPtr<WebKit::WebViewImpl>(_webView), remainingID = data.unanimatedRangeUUID, uuid = RetainPtr(uuid)] {
            auto strongWebView = weakWebView.get();
            if (strongWebView)
                strongWebView->page().updateUnderlyingTextVisibilityForTextAnimationID(remainingID, true);
            auto animationID = WTF::UUID::fromNSUUID(uuid.get());
            if (animationID)
                strongWebView->page().callCompletionHandlerForAnimationID(*animationID);
```

File: Source/WebKit/WebProcess/WebPage/Cocoa/TextAnimationController.mm
Comment (pxlcoder) [inline Source/WebKit/WebProcess/WebPage/Cocoa/TextAnimationController.mm:217]:
```suggestion
        if (previouslyReplacedRange->range == replacingRange && previouslyReplacedRange->string == string)
```

Comment (pxlcoder) [reply Source/WebKit/WebProcess/WebPage/Cocoa/TextAnimationController.mm:217]: May even be nice to pass in a `ReplacedRangeAndString` as the parameter, so you can directly compare to that, but I won't hold this PR up for that.

File: Source/WebKit/WebProcess/WebPage/Cocoa/TextAnimationController.mm
Comment (pxlcoder) [inline Source/WebKit/WebProcess/WebPage/Cocoa/TextAnimationController.mm:220]:
```suggestion
        replaceCharacterRange = remainingCharacterRange(replacingRange, previouslyReplacedRange->range);
```

File: Tools/TestWebKitAPI/Tests/WebKitCocoa/WritingTools.mm
Comment (pxlcoder) [inline Tools/TestWebKitAPI/Tests/WebKitCocoa/WritingTools.mm:818]: (for my own understanding: this is just for ease of debugging right?)

File: Tools/TestWebKitAPI/Tests/WebKitCocoa/WritingTools.mm
Comment (pxlcoder) [inline Tools/TestWebKitAPI/Tests/WebKitCocoa/WritingTools.mm:1249]:
```suggestion
        TestWebKitAPI::Util::runFor(0.1_s); // This is needed to let the information load at least a bit or we can fail on a JavaScript issue.
```

Comment (pxlcoder) [reply Tools/TestWebKitAPI/Tests/WebKitCocoa/WritingTools.mm:1249]: I'm curious what the JS issue is here? Given `waitForValue` already waits for the image to load.

File: Tools/TestWebKitAPI/Tests/WebKitCocoa/WritingTools.mm
Comment (pxlcoder) [inline Tools/TestWebKitAPI/Tests/WebKitCocoa/WritingTools.mm]: Unnecessary change.

File: Source/WebKit/UIProcess/TextAnimationType.h
Comment (pxlcoder) [inline Source/WebKit/UIProcess/TextAnimationType.h:40]: Is `{ WTF::UUID::emptyValue }` really necessary?

**Review by pxlcoder — APPROVED** [review]
Slightly curious how long the test suite takes to run now, given all the waiting. 

(I understand it is necessary due to the nature of the animations, just curious)

File: Tools/TestWebKitAPI/Tests/WebKitCocoa/WritingTools.mm
Comment (megangardner) [reply Tools/TestWebKitAPI/Tests/WebKitCocoa/WritingTools.mm:818]: yes, I like the more 'reasonable' strings, but then I couldn't tell which order they were supposed to come in, so I added suffixes.

File: Source/WebKit/WebProcess/WebPage/Cocoa/TextAnimationController.mm
Comment (megangardner) [reply Source/WebKit/WebProcess/WebPage/Cocoa/TextAnimationController.mm:217]: So, the reason this is (*previouslyReplacedRange)->range and not previouslyReplacedRange->range is because this is a double-wrapped optional. yes, I know it's crazy, I would like to fix it, but this is where we are right now.

File: Source/WebKit/WebProcess/WebPage/Cocoa/TextAnimationController.mm
Comment (megangardner) [reply Source/WebKit/WebProcess/WebPage/Cocoa/TextAnimationController.mm:220]: ditto

File: Source/WebKit/UIProcess/TextAnimationType.h
Comment (megangardner) [reply Source/WebKit/UIProcess/TextAnimationType.h:40]: TextAnimationType.h:40:35: error: field of type 'WTF::UUID' has private default constructor
   40 |     WTF::UUID unanimatedRangeUUID { };

So, yes.

File: Tools/TestWebKitAPI/Tests/WebKitCocoa/WritingTools.mm
Comment (megangardner) [reply Tools/TestWebKitAPI/Tests/WebKitCocoa/WritingTools.mm:1249]: Encountered error: Error Domain=WKErrorDomain Code=4 "A JavaScript exception occurred" UserInfo={WKJavaScriptExceptionLineNumber=1, WKJavaScriptExceptionMessage=TypeError: null is not an object (evaluating 'document.querySelector('img').complete'), WKJavaScriptExceptionColumnNumber=30, WKJavaScriptExceptionSourceURL=file:///..../TestWebKitAPI.resources/, NSLocalizedDescription=A JavaScript exception occurred} while evaluating script: document.querySelector('img').complete

Comment (pxlcoder) [reply Tools/TestWebKitAPI/Tests/WebKitCocoa/WritingTools.mm:1249]: Oh, the `` element wasn't created. It might be nicer to update the `waitForValue` to also wait for that to happen.

**Comment by megangardner** [comment]
> Slightly curious how long the test suite takes to run now, given all the waiting.
> 
> (I understand it is necessary due to the nature of the animations, just curious)

About 90 seconds

**Pull request merged**
